File: Solutions/1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.js
/**
 * @param {number} n
 * @param {number} k
 * @return {string}
 */
var getHappyString = function(n, k) {
    let result = [];
    
    // Helper function to generate the happy strings
    function dfs(current) {
        // If the length of the current string is equal to n, add it to result
        if (current.length === n) {
            result.push(current);
            return;
        }
        
        // Try all three characters 'a', 'b', 'c'
        for (let char of ['a', 'b', 'c']) {
            // Ensure the current character is not the same as the previous one
            if (current[current.length - 1] !== char) {
                dfs(current + char);
            }
        }
    }

    // Start DFS with an empty string
    dfs('');
    
    // If there are enough strings, return the k-th string (1-indexed)
    return k <= result.length ? result[k - 1] : "";
};

File: Solutions/1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.dart
class Solution {
  String getHappyString(int n, int k) {
    List<String> result = [];
    List<String> letters = ['a', 'b', 'c'];

    // Backtracking function to build happy strings
    void backtrack(String current) {
      // If we already have k strings, stop early
      if (result.length >= k) return;

      // If the current string reaches length n, add it to result
      if (current.length == n) {
        result.add(current);
        return;
      }

      for (String ch in letters) {
        // Skip if same as previous character (to maintain happy string)
        if (current.isNotEmpty && current[current.length - 1] == ch) continue;

        backtrack(current + ch); // Recurse with new character
      }
    }

    backtrack(''); // Start with empty string

    // Return k-th string if it exists, else empty string
    return k <= result.length ? result[k - 1] : '';
  }
}
